pkg/tools/dnsutils/memory/handler.go

```diff
@@ -72,18 +68,23 @@ func (f *memoryHandler) ServeDNS(ctx context.Context, rw dns.ResponseWriter, msg
 		resp.Answer = append(resp.Answer, f.ptr(name)...)
 	}
 
-	if len(resp.Answer) == 0 {
-		next.Handler(ctx).ServeDNS(ctx, rwWrapper, msg)
-		if !rwWrapper.passed {
-			// Send NXDomain because we didn't find anything
-			m := new(dns.Msg)
-			_ = rw.WriteMsg(m.SetRcode(msg, dns.RcodeNameError))
+	if len(resp.Answer) != 0 {
+		if err := rw.WriteMsg(resp); err != nil {
+			dns.HandleFailed(rw, msg)
 		}
 		return
 	}
 
-	if err := rw.WriteMsg(resp); err != nil {
-		dns.HandleFailed(rw, msg)
+	if _, ok := f.records.Load(name); ok {
+		m := new(dns.Msg)
+		_ = rw.WriteMsg(m.SetRcode(msg, dns.RcodeSuccess))
+	} else {
+		rwWrapper := &responseWriter{ResponseWriter: rw}
```

Comment:
1. please remove `responseWriter.`

`fanout` should handler the server fail.

2. add regression unit test

Reply:
`responseWriter` doesn't check only fanout, but any handler that goes after memory handler. If we remove `responseWriter` we can get an error, because in some cases nothing will be written to `rw`.